### PR TITLE
Add basic Flutter Preview view.

### PR DESCRIPTION
**Pull request by scheglov**
1. Add `FlutterDartAnalysisServer` with support for Flutter specific subscriptions.
2. Add subscription and listener for `OUTLINE` service.
3. Add a new `Flutter Preview` view in IntelliJ and show Flutter outline in it.
4. The view shows Dart elements and Flutter widgets.
5. Parent/child association, class, (if a variable) variable name, and attributes are shown for widgets.
6. Double click on a widget for navigation.

What does not work.
1. If you start with the view open, and Analysis Server Service is not ready yet, you probably will get an exception.
2. If you restart the Analysis Server, you will lose your subscription and will not get updates.
3. No navigation from source.

### CODE REVIEW

**Comment by scheglov** [comment]
image

**Comment by devoncarew** [comment]
Awesome! Will review in a bit; one first thought is that we should add this work behind a flag - give ourselves some time to iterate and get feedback - and then remove the flag. We had something similar here: https://github.com/flutter/flutter-intellij/pull/1619/files (that PR is where we removed the experimental flags section - we'd want to add that back in for this work).

File: resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -536,6 +536,13 @@
     <annotator language="Dart" implementationClass="io.flutter.editor.FlutterEditorAnnotator"/>
     <errorHandler implementation="io.flutter.FlutterErrorReportSubmitter"/>
 
+    <toolWindow id="Flutter Preview" anchor="right" icon="FlutterIcons.Flutter_13"
+                factoryClass="io.flutter.view.preview.PreviewViewFactory"
+                conditionClass="io.flutter.view.FlutterViewCondition"/>
```
Comment (devoncarew) [inline resources/META-INF/plugin.xml]: Let's switch this to a new `FlutterPreviewCondition`, and have that also check the state of the preference to enable 'Flutter Preview'.

File: resources/META-INF/plugin.xml [hunk 2]
```diff
@@ -536,6 +536,13 @@
     <annotator language="Dart" implementationClass="io.flutter.editor.FlutterEditorAnnotator"/>
     <errorHandler implementation="io.flutter.FlutterErrorReportSubmitter"/>
 
+    <toolWindow id="Flutter Preview" anchor="right" icon="FlutterIcons.Flutter_13"
```
Comment (devoncarew) [inline resources/META-INF/plugin.xml]: @jwren brings up a good point, that we should iterate on the name (not for this PR). `Flutter preview` as a name may not be obvious to users what it is; others candidates are `Flutter Outline`, `Flutter Structure` (familiar to existing IntelliJ users), and `Flutter Designer`.

We're good with the current name but should keep in the back of our minds that we'll need to settle on what we want the final name to be.

File: src/io/flutter/view/preview/DartElementPresentationUtil.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  public static Icon getIcon(Element element) {
+    final boolean finalOrConst = element.isConst() || element.isFinal();
+
+    switch (element.getKind()) {
+      case ElementKind.CLASS:
+        return element.isAbstract() ? AbstractClass : Class;
+      case ElementKind.CONSTRUCTOR:
+        return Method;
+      //case ElementKind.CONSTRUCTOR_INVOCATION:
```
Comment (devoncarew) [inline src/io/flutter/view/preview/DartElementPresentationUtil.java]: This is not yet available in the earliest Dart plugin we need to support? If so, can you make a TODO: for us to use this once the min. version revs to 2017.3 (or whichever is appropriate)?

File: src/io/flutter/view/preview/DartElementPresentationUtil.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      case ElementKind.ENUM_CONSTANT:
+        return STATIC_FINAL_FIELD_ICON;
+      case ElementKind.FIELD:
+        if (finalOrConst && element.isTopLevelOrStatic()) return STATIC_FINAL_FIELD_ICON;
+        if (finalOrConst) return FINAL_FIELD_ICON;
+        if (element.isTopLevelOrStatic()) return STATIC_FIELD_ICON;
+        return Field;
+      case ElementKind.FUNCTION:
+        return element.isTopLevelOrStatic() ? TOP_LEVEL_FUNCTION_ICON : Function;
+      //case ElementKind.FUNCTION_INVOCATION:
```
Comment (devoncarew) [inline src/io/flutter/view/preview/DartElementPresentationUtil.java]: same

File: src/io/flutter/view/preview/DartElementPresentationUtil.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  private static final LayeredIcon FINAL_FIELD_ICON = new LayeredIcon(Field, FinalMark);
+  private static final LayeredIcon STATIC_FIELD_ICON = new LayeredIcon(Field, StaticMark);
+  private static final LayeredIcon STATIC_METHOD_ICON = new LayeredIcon(Method, StaticMark);
+  private static final LayeredIcon TOP_LEVEL_FUNCTION_ICON = new LayeredIcon(Function, StaticMark);
+  private static final LayeredIcon TOP_LEVEL_VAR_ICON = new LayeredIcon(Variable, StaticMark);
+  private static final LayeredIcon CONSTRUCTOR_INVOCATION_ICON = new LayeredIcon(Class, TabPin);
+  private static final LayeredIcon FUNCTION_INVOCATION_ICON = new LayeredIcon(Method, TabPin);
+  private static final LayeredIcon TOP_LEVEL_CONST_ICON = new LayeredIcon(Variable, StaticMark, FinalMark);
+
+  public static Icon getIcon(Element element) {
```
Comment (devoncarew) [inline src/io/flutter/view/preview/DartElementPresentationUtil.java]: Is this from the Dar plugin, and if so, can you add a link to the file on github (to help us better keep them in sync)?

File: src/io/flutter/view/preview/PreviewView.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    final DefaultMutableTreeNode rootNode = new DefaultMutableTreeNode();
+
+    tree = new OutlineTree(rootNode);
+    tree.setCellRenderer(new OutlineTreeCellRenderer());
+    tree.expandAll();
+
+    tree.addMouseListener(new MouseAdapter() {
+      @Override
+      public void mouseClicked(MouseEvent e) {
+        if (e.getClickCount() == 2) {
```
Comment (devoncarew) [inline src/io/flutter/view/preview/PreviewView.java]: Is there no double click accessor?

File: resources/META-INF/plugin.xml (hunk 1 above)
Comment (scheglov) [reply resources/META-INF/plugin.xml]: Done.

File: src/io/flutter/view/preview/DartElementPresentationUtil.java (hunk 5 above)
Comment (scheglov) [reply src/io/flutter/view/preview/DartElementPresentationUtil.java]: I added the link to the class documentation.

File: src/io/flutter/view/preview/DartElementPresentationUtil.java (hunk 3 above)
Comment (scheglov) [reply src/io/flutter/view/preview/DartElementPresentationUtil.java]: Yes, it is `2017.3`, I added comments here and below.

File: src/io/flutter/view/preview/PreviewView.java (hunk 6 above)
Comment (scheglov) [reply src/io/flutter/view/preview/PreviewView.java]: No, I think this is how it is done in Swing.
On the bright side, we can also have triple clicks :-)

**Comment by scheglov** [comment]
PTAL

**Comment by devoncarew** [comment]
👍 lgtm

**Pull request merged**